langchain4j/src/test/java/dev/langchain4j/service/common/AbstractAiServiceWithToolsIT.java

```
@@ -785,4 +810,338 @@ protected void should_execute_tool_with_uuid_parameter(ChatModel model) {
+        assertThat(response.toolExecutions().get(0).result()).isEqualTo("124");
+
+        List<ChatMessage> messages = chatMemory.messages();
+        assertThat(messages).hasSize(3);
+        checkMemoryWithImmediateTool(messages, "How much is 37 plus 87?", "124");
+
+        // Check that the memory is not corrupted and conversation can continue
+        response = assistant.chat("Now add 47 to the previous result");
+        assertThat(response.content()).isNull();
+        assertThat(response.toolExecutions().get(0).result()).isEqualTo("171");
+
+        messages = chatMemory.messages();
+        assertThat(messages).hasSize(6);
+        checkMemoryWithImmediateTool(messages.subList(3, 6), "Now add 47 to the previous result", "171");
+    }
+
+    static void checkMemoryWithImmediateTool(List<ChatMessage> messages, String request, String response) {
+        // The first message is the user request
+        assertThat(messages.get(0)).isInstanceOf(UserMessage.class);
+        assertThat(((UserMessage) messages.get(0)).singleText()).isEqualTo(request);
+
+        // The second message is the AI response with tool execution request
+        assertThat(messages.get(1)).isInstanceOf(AiMessage.class);
+        assertThat(((AiMessage) messages.get(1)).toolExecutionRequests()).hasSize(1);
+        assertThat(((AiMessage) messages.get(1)).toolExecutionRequests().get(0).name()).isEqualTo("add");
+
+        // The third message is the tool execution result
+        assertThat(messages.get(2)).isInstanceOf(ToolExecutionResultMessage.class);
+        assertThat(((ToolExecutionResultMessage) messages.get(2)).toolName()).isEqualTo("add");
+        assertThat(((ToolExecutionResultMessage) messages.get(2)).text()).isEqualTo(response);
+    }
+
+    interface StringAssistant {
+        String chat(String userMessage);
+    }
+
+    @ParameterizedTest
+    @MethodSource("models")
+    protected void should_ignore_immediate_tool_when_not_returning_Result(ChatModel model) {
```

I would rather throw an exception as soon as possible if AI service without `Result<T>` return type is configured with at least one immediate tool, WDYT? Otherwise user might be confused for a long time why it does not work

---

Done.